src/main/java/graphql/execution/ValuesResolver.java

```diff
@@ -373,12 +375,42 @@ private static Map<String, Object> getArgumentValuesImpl(
                             locale);
                     coercedValues.put(argumentName, value);
                 }
+                // @oneOf input must be checked now that all variables and literals have been converted
+                if (argumentType instanceof GraphQLInputObjectType) {
+                    GraphQLInputObjectType inputObjectType = (GraphQLInputObjectType) argumentType;
+                    if (inputObjectType.isOneOf() && ! ValuesResolverConversion.isNullValue(value)) {
+                        validateOneOfInputTypes(inputObjectType, argumentValue, argumentName, value, locale);
+                    }
+                }
             }
 
         }
         return coercedValues;
     }
 
+    @SuppressWarnings("unchecked")
+    private static void validateOneOfInputTypes(GraphQLInputObjectType oneOfInputType, Value argumentValue, String argumentName, Object inputValue, Locale locale) {
+        Assert.assertTrue(inputValue instanceof Map, () -> String.format("The coerced argument %s GraphQLInputObjectType is unexpectedly not a map", argumentName));
+        Map<String, Object> objectMap = (Map<String, Object>) inputValue;
+        int mapSize;
+        if (argumentValue instanceof ObjectValue) {
+            mapSize = ((ObjectValue) argumentValue).getObjectFields().size();
+        } else {
+            mapSize = objectMap.size();
+        }
+        if (mapSize != 1) {
+            String msg = I18n.i18n(I18n.BundleType.Execution, locale)
+                    .msg("Execution.handleOneOfNotOneFieldError", oneOfInputType.getName());
+            throw new OneOfTooManyKeysException(msg);
+        }
+        String fieldName = objectMap.keySet().iterator().next();
+        if (objectMap.get(fieldName) == null) {
+            String msg = I18n.i18n(I18n.BundleType.Execution, locale)
+                    .msg("Execution.handleOneOfValueIsNullError", oneOfInputType.getName() + "." + fieldName);
+            throw new OneOfNullValueException(msg);
+        }
+    }
+
```

Comment:
remember we can have a mix of input - literals and maps and invalid input might be a with a variable with no map entries or multiple.

The unit tests for this explore these different conditions